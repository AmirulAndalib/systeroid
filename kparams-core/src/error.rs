use thiserror::Error as ThisError;

/// Custom error type.
#[derive(Debug, ThisError)]
pub enum Error {
    /// Error that might occur during I/O operations.
    #[error("IO error: `{0}`")]
    IoError(#[from] std::io::Error),
    /// Error that might occur whenever a lock is acquired.
    #[error("Thread lock error: `{0}`")]
    ThreadLockError(String),
    /// Error that might occur while parsing documents.
    #[error("parse error: `{0}`")]
    ParseError(String),
}

unsafe impl Send for Error {}

/// Type alias for the standard [`Result`] type.
pub type Result<T> = core::result::Result<T, Error>;
