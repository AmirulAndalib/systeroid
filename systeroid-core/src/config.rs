use crate::sysctl::Section;
use colored::Color;
use std::collections::HashMap;

/* Macro for the concise initialization of HashMap */
macro_rules! map {
    ($( $key: expr => $val: expr ),*) => {{
         let mut map = ::std::collections::HashMap::new();
         $( map.insert($key, $val); )*
         map
    }}
}

/// General configuration.
#[derive(Debug, Default)]
pub struct Config {
    /// Sysctl configuration.
    pub sysctl: SysctlConfig,
}

/// Sysctl configuration.
#[derive(Debug)]
pub struct SysctlConfig {
    /// Sections and the corresponding colors.
    pub section_colors: HashMap<Section, Color>,
    /// Default color for the output
    pub default_color: Color,
}

impl Default for SysctlConfig {
    fn default() -> Self {
        Self {
            section_colors: map! {
                Section::Abi => Color::Red,
                Section::Fs => Color::Green,
                Section::Kernel => Color::Magenta,
                Section::Net => Color::Blue,
                Section::Sunrpc => Color::Yellow,
                Section::User => Color::Cyan,
                Section::Vm => Color::BrightRed,
                Section::Unknown => Color::BrightBlack
            },
            default_color: Color::BrightBlack,
        }
    }
}
