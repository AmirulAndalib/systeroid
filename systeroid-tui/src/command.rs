use std::str::FromStr;
use termion::event::Key;

/// Possible application commands.
#[derive(Debug)]
pub enum Command {
    /// Perform an action based on the selected entry.
    Select,
    /// Scroll up on the widget.
    ScrollUp,
    /// Scroll down on the widget.
    ScrollDown,
    /// Move cursor to right/left.
    MoveCursor(u8),
    /// Enable the search mode.
    EnableSearch,
    /// Process the input.
    ProcessInput,
    /// Update the input buffer.
    UpdateInput(char),
    /// Clear the input buffer.
    ClearInput(bool),
    /// Copy selected value to clipboard.
    Copy,
    /// Refresh the application.
    Refresh,
    /// Exit the application.
    Exit,
    /// Do nothing.
    None,
}

impl FromStr for Command {
    type Err = ();
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "search" => Ok(Command::EnableSearch),
            "up" => Ok(Command::ScrollUp),
            "down" => Ok(Command::ScrollDown),
            "copy" => Ok(Command::Copy),
            "refresh" => Ok(Command::Refresh),
            "exit" | "quit" | "q" | "q!" => Ok(Command::Exit),
            _ => Err(()),
        }
    }
}

impl Command {
    /// Parses a command from the given key.
    pub fn parse(key: Key, input_mode: bool) -> Self {
        if input_mode {
            match key {
                Key::Char('\n') => Command::ProcessInput,
                Key::Char(c) => Command::UpdateInput(c),
                Key::Backspace => Command::ClearInput(false),
                Key::Delete => Command::ClearInput(true),
                Key::Left => Command::MoveCursor(1),
                Key::Right => Command::MoveCursor(0),
                Key::Esc => Command::Exit,
                _ => Command::None,
            }
        } else {
            match key {
                Key::Up => Command::ScrollUp,
                Key::Down => Command::ScrollDown,
                Key::Char(':') => Command::UpdateInput(' '),
                Key::Char('/') => Command::EnableSearch,
                Key::Char('\n') => Command::Select,
                Key::Char('c') => Command::Copy,
                Key::Char('r') => Command::Refresh,
                Key::Esc => Command::Exit,
                _ => Command::None,
            }
        }
    }
}
