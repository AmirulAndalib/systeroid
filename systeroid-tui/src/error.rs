use thiserror::Error as ThisError;

/// Custom error type.
#[derive(Debug, ThisError)]
pub enum Error {
    /// Error that may occur during I/O operations.
    #[error("IO error: `{0}`")]
    IoError(#[from] std::io::Error),
    /// Error that may occur while receiving messages from the channel.
    #[error("Channel receive error: `{0}`")]
    ReceiveError(#[from] std::sync::mpsc::RecvError),
}

/// Type alias for the standard [`Result`] type.
pub type Result<T> = core::result::Result<T, Error>;
