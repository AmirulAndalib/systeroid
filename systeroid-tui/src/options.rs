use std::convert::TryFrom;

/// Available copying options.
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum CopyOption {
    /// Copy the name of the parameter.
    Name,
    /// Copy the value of the parameter.
    Value,
    /// Copy the documentation of the parameter.
    Documentation,
}

impl<'a> TryFrom<&'a str> for CopyOption {
    type Error = ();
    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
        Self::variants()
            .iter()
            .find(|v| value == v.as_str())
            .copied()
            .ok_or(())
    }
}

impl CopyOption {
    /// Returns the string representation of the option.
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Name => "Parameter name",
            Self::Value => "Parameter value",
            Self::Documentation => "Documentation",
        }
    }

    /// Returns the variants.
    pub fn variants() -> &'static [Self] {
        &[Self::Name, Self::Value, Self::Documentation]
    }
}
