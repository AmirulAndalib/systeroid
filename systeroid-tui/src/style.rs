use crate::error::Result;
use std::str::FromStr;
use tui::style::{Color as TuiColor, Style};

/// Color configuration.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Colors {
    /// Background color.
    bg: TuiColor,
    /// Foreground color.
    fg: TuiColor,
}

impl Default for Colors {
    fn default() -> Self {
        Colors {
            bg: TuiColor::Reset,
            fg: TuiColor::Reset,
        }
    }
}

impl Colors {
    /// Constructs a new instance.
    pub fn new(background: &str, foreground: &str) -> Result<Self> {
        Ok(Self {
            bg: TuiColor::from_str(background)?,
            fg: TuiColor::from_str(foreground)?,
        })
    }

    /// Returns the background/foreground colors with default style.
    pub fn get_style(&self) -> Style {
        Style::default().bg(self.fg).fg(self.bg)
    }

    /// Returns the background color with default style.
    pub fn get_bg_style(&self) -> Style {
        Style::default().bg(self.bg)
    }

    /// Returns the foreground color with default style.
    pub fn get_fg_style(&self) -> Style {
        Style::default().fg(self.fg)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_style() -> Result<()> {
        assert_eq!(
            Colors {
                bg: TuiColor::from_str("red")?,
                fg: TuiColor::from_str("blue")?,
            },
            Colors::new("red", "blue")?
        );
        assert_eq!(
            Style::default().fg(TuiColor::Green),
            Colors::new("reset", "Green")?.get_fg_style()
        );
        assert_eq!(
            Style::default().bg(TuiColor::Yellow),
            Colors::new("YELLOW", "reset")?.get_bg_style()
        );
        assert_eq!(
            Style::default()
                .bg(TuiColor::DarkGray)
                .fg(TuiColor::Magenta),
            Colors::new("Magenta", "DarkGray")?.get_style()
        );
        Ok(())
    }
}
