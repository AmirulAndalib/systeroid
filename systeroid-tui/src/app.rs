use crate::command::Command;
use crate::error::Result;
use crate::widgets::StatefulTable;
use std::str::FromStr;
use std::time::Instant;
use systeroid_core::sysctl::controller::Sysctl;
use systeroid_core::sysctl::parameter::Parameter;

/// Duration of prompt messages.
const MESSAGE_DURATION: u128 = 1750;

/// Application controller.
#[derive(Debug)]
pub struct App<'a> {
    /// Whether if the application is running.
    pub running: bool,
    /// Input buffer.
    pub input: Option<String>,
    /// Time tracker for measuring the time for clearing the input.
    pub input_time: Option<Instant>,
    /// List of sysctl parameters.
    pub parameter_list: StatefulTable<Parameter>,
    /// Sysctl controller.
    sysctl: &'a mut Sysctl,
}

impl<'a> App<'a> {
    /// Constructs a new instance.
    pub fn new(sysctl: &'a mut Sysctl) -> Self {
        Self {
            running: true,
            input: None,
            input_time: None,
            parameter_list: StatefulTable::with_items(sysctl.parameters.clone()),
            sysctl,
        }
    }

    /// Returns true if the app is in input mode.
    pub fn is_input_mode(&self) -> bool {
        self.input.is_some() && self.input_time.is_none()
    }

    /// Runs the given command and updates the application.
    pub fn run_command(&mut self, command: Command) -> Result<()> {
        match command {
            Command::ScrollUp => {
                self.parameter_list.previous();
            }
            Command::ScrollDown => {
                self.parameter_list.next();
            }
            Command::ProcessInput => {
                if self.input_time.is_some() {
                    return Ok(());
                } else if let Some(input) = &self.input {
                    if let Ok(command) = Command::from_str(input) {
                        self.run_command(command)?;
                    } else {
                        self.input = Some(String::from("Unknown command"));
                        self.input_time = Some(Instant::now());
                    }
                }
            }
            Command::UpdateInput(v) => match self.input.as_mut() {
                Some(input) => {
                    if self.input_time.is_some() {
                        self.input_time = None;
                        self.input = Some(String::new());
                    } else {
                        input.push(v);
                    }
                }
                None => {
                    self.input = Some(String::new());
                }
            },
            Command::ClearInput(cancel) => {
                if self.input_time.is_some() {
                    return Ok(());
                } else if cancel {
                    self.input = None
                } else if let Some(input) = self.input.as_mut() {
                    if input.pop().is_none() {
                        self.input = None;
                    }
                }
            }
            Command::Refresh => {
                self.input = None;
                *self.sysctl = Sysctl::init(self.sysctl.config.clone())?;
                self.parameter_list = StatefulTable::with_items(self.sysctl.parameters.clone());
            }
            Command::Exit => {
                self.running = false;
            }
            Command::None => {}
        }
        Ok(())
    }

    /// Handles the terminal tick event.
    pub fn tick(&mut self) {
        if let Some(instant) = self.input_time {
            if instant.elapsed().as_millis() > MESSAGE_DURATION {
                self.input = None;
                self.input_time = None;
            }
        }
    }
}
