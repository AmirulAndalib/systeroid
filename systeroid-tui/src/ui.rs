use crate::app::App;
use tui::backend::Backend;
use tui::layout::{Constraint, Direction, Layout, Rect};
use tui::style::{Color, Style};
use tui::widgets::{Block, BorderType, Borders, Cell, Paragraph, Row, Table};
use tui::Frame;
use unicode_width::UnicodeWidthStr;

/// Renders the user interface.
pub fn render<B: Backend>(frame: &mut Frame<'_, B>, app: &mut App) {
    let rect = frame.size();
    let chunks = Layout::default()
        .direction(Direction::Horizontal)
        .constraints([Constraint::Percentage(50), Constraint::Percentage(50)].as_ref())
        .split(rect);
    {
        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints([Constraint::Min(rect.height - 3), Constraint::Min(3)].as_ref())
            .split(chunks[0]);
        render_parameter_list(frame, chunks[0], app);
        render_input_prompt(frame, chunks[1], rect.height - 2, app);
    }
    render_parameter_documentation(frame, chunks[1], app);
}

/// Renders the list that contains the sysctl parameters.
fn render_parameter_list<B: Backend>(frame: &mut Frame<'_, B>, rect: Rect, app: &mut App) {
    let max_width = app
        .parameter_list
        .items
        .iter()
        .map(|p| p.name.len())
        .max_by(|x, y| x.cmp(y))
        .and_then(|v| u16::try_from(v).ok())
        .unwrap_or(1);
    let minimize_rows = rect.width < max_width + 10;
    let rows = app.parameter_list.items.iter().map(|item| {
        Row::new(if minimize_rows {
            vec![Cell::from(format!("{} = {}", item.name, item.value))]
        } else {
            vec![
                Cell::from(item.name.clone()),
                Cell::from(item.value.clone()),
            ]
        })
        .height(1)
        .bottom_margin(1)
    });
    frame.render_stateful_widget(
        Table::new(rows)
            .block(
                Block::default()
                    .borders(Borders::all())
                    .border_style(Style::default().fg(Color::White))
                    .border_type(BorderType::Rounded)
                    .style(Style::default().bg(Color::Black)),
            )
            .highlight_style(Style::default().bg(Color::White).fg(Color::Black))
            .widths(&if minimize_rows {
                [Constraint::Percentage(100), Constraint::Min(0)]
            } else {
                [Constraint::Min(max_width), Constraint::Percentage(100)]
            }),
        rect,
        &mut app.parameter_list.state,
    );
}

/// Renders the documentation of the selected sysctl parameter.
fn render_parameter_documentation<B: Backend>(frame: &mut Frame<'_, B>, rect: Rect, app: &mut App) {
    frame.render_widget(
        Paragraph::new(
            app.parameter_list
                .selected()
                .and_then(|parameter| parameter.get_documentation())
                .unwrap_or_else(|| String::from("No documentation available")),
        )
        .block(
            Block::default()
                .borders(Borders::all())
                .border_style(Style::default().fg(Color::White))
                .border_type(BorderType::Rounded)
                .style(Style::default().bg(Color::Black)),
        ),
        rect,
    );
}

/// Renders the input prompt for running commands.
fn render_input_prompt<B: Backend>(
    frame: &mut Frame<'_, B>,
    rect: Rect,
    cursor_y: u16,
    app: &mut App,
) {
    let text = match &app.input {
        Some(input) => format!(
            "{}{}",
            if app.input_time.is_some() {
                "MSG: "
            } else {
                frame.set_cursor(input.width() as u16 + 2, cursor_y);
                ":"
            },
            input,
        ),
        None => String::new(),
    };
    frame.render_widget(
        Paragraph::new(text).block(
            Block::default()
                .borders(Borders::all())
                .border_style(Style::default().fg(Color::White))
                .border_type(BorderType::Rounded)
                .style(Style::default().bg(Color::Black)),
        ),
        rect,
    );
}
