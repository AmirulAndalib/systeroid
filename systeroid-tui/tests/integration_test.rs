use insta::assert_snapshot;
use ratatui::backend::TestBackend;
use ratatui::Terminal;
use std::path::PathBuf;
use std::thread;
use std::time::Duration;
use systeroid_core::config::Config;
use systeroid_core::sysctl::controller::Sysctl;
use systeroid_core::sysctl::parameter::Parameter;
use systeroid_core::sysctl::section::Section;
use systeroid_tui::app::App;
use systeroid_tui::command::Command;
use systeroid_tui::error::Result;
use systeroid_tui::options::{Direction, ScrollArea};
use systeroid_tui::style::Colors;
use systeroid_tui::ui::render;

#[test]
fn test_render_tui() -> Result<()> {
    let mut sysctl = Sysctl {
        parameters: vec![
            Parameter {
                name: String::from("user.name"),
                value: String::from("system"),
                description: None,
                section: Section::User,
                docs_path: PathBuf::new(),
                docs_title: String::new(),
            },
            Parameter {
                name: String::from("kernel.fictional.test_param"),
                value: String::from("0"),
                description: Some(String::from("This is a fictional parameter for testing")),
                section: Section::Kernel,
                docs_path: PathBuf::from("/etc/cosmos"),
                docs_title: String::from("Test Parameter"),
            },
            Parameter {
                name: String::from("vm.stat_interval"),
                value: String::from("1"),
                description: Some(String::from(
                    "The time interval between which vm statistics are updated",
                )),
                section: Section::Vm,
                docs_path: PathBuf::from("/usr/share/doc/linux/admin-guide/sysctl/vm.rst"),
                docs_title: String::from("stat_interval"),
            },
        ],
        config: Config::default(),
    };
    let mut app = App::new(&mut sysctl);
    let colors = Colors::default();
    let backend = TestBackend::new(40, 10);
    let mut terminal = Terminal::new(backend)?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("default", terminal.backend());

    app.run_command(Command::Help)?;
    app.run_command(Command::Scroll(ScrollArea::List, Direction::Down, 1))?;
    app.run_command(Command::Scroll(ScrollArea::List, Direction::Up, 1))?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("about", terminal.backend());
    app.run_command(Command::Select)?;

    app.run_command(Command::Select)?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("set", terminal.backend());
    assert!(app.is_input_mode());

    app.run_command(Command::ClearInput(false))?;
    app.run_command(Command::MoveCursor(Direction::Left))?;
    app.run_command(Command::ClearInput(true))?;
    app.run_command(Command::ClearInput(true))?;
    "kill"
        .chars()
        .try_for_each(|c| app.run_command(Command::UpdateInput(c)))?;

    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("input", terminal.backend());

    app.run_command(Command::ProcessInput)?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("process_input", terminal.backend());

    thread::sleep(Duration::from_millis(2000));
    app.tick();
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("go_back", terminal.backend());

    app.run_command(Command::Search)?;
    app.run_command(Command::Cancel)?;
    app.run_command(Command::Copy)?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("copy", terminal.backend());

    app.run_command(Command::Scroll(ScrollArea::List, Direction::Down, 1))?;
    app.run_command(Command::Scroll(ScrollArea::List, Direction::Up, 1))?;
    app.run_command(Command::Select)?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("copy_not_enabled", terminal.backend());

    thread::sleep(Duration::from_millis(2000));
    app.tick();
    app.run_command(Command::Scroll(ScrollArea::Section, Direction::Left, 1))?;
    app.run_command(Command::Scroll(ScrollArea::Section, Direction::Left, 1))?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("list", terminal.backend());
    app.run_command(Command::Scroll(ScrollArea::Section, Direction::Right, 1))?;
    app.run_command(Command::Scroll(ScrollArea::Section, Direction::Right, 1))?;

    app.input = Some(String::new());
    app.run_command(Command::Search)?;
    app.run_command(Command::UpdateInput('_'))?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("search", terminal.backend());

    app.run_command(Command::ProcessInput)?;
    app.run_command(Command::Scroll(ScrollArea::Documentation, Direction::Up, 1))?;
    app.run_command(Command::Scroll(
        ScrollArea::Documentation,
        Direction::Down,
        5,
    ))?;
    app.run_command(Command::Scroll(ScrollArea::Documentation, Direction::Up, 1))?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("documentation", terminal.backend());

    app.run_command(Command::Scroll(ScrollArea::List, Direction::Bottom, 1))?;
    app.run_command(Command::Scroll(ScrollArea::List, Direction::Up, 1))?;
    app.run_command(Command::Scroll(ScrollArea::List, Direction::Up, 2))?;
    app.run_command(Command::Scroll(ScrollArea::List, Direction::Top, 1))?;
    app.run_command(Command::Scroll(ScrollArea::List, Direction::Down, 1))?;
    app.run_command(Command::Scroll(ScrollArea::List, Direction::Down, 2))?;
    app.run_command(Command::Refresh)?;
    terminal.draw(|frame| render(frame, &mut app, &colors))?;
    assert_snapshot!("refreshed", terminal.backend());

    app.run_command(Command::Nothing)?;
    app.run_command(Command::Exit)?;
    assert!(!app.running);

    Ok(())
}
