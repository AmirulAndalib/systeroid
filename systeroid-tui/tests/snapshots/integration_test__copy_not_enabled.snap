---
source: systeroid-tui/tests/integration_test.rs
expression: terminal.backend()
---
"╭Parameters──────────────────────|all|─╮"
"│user.name                   system    │"
"│kernel.fictional.test_param 0         │"
"│vm.stat_interval            1         │"
"│                                      │"
"│                                  1/3 │"
"╰──────────────────────────────────────╯"
"╭──────────────────────────────────────╮"
"│MSG: Clipboard support is not enabled │"
"╰──────────────────────────────────────╯"
